sdk_extensions/logging/support/src/main/java/io/opentelemetry/logging/BatchLogProcessor.java

```diff
@@ -0,0 +1,372 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.logging;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.common.Labels;
+import io.opentelemetry.internal.Utils;
+import io.opentelemetry.logging.api.Exporter;
+import io.opentelemetry.logging.api.LogProcessor;
+import io.opentelemetry.logging.api.LogRecord;
+import io.opentelemetry.metrics.LongCounter;
+import io.opentelemetry.metrics.LongCounter.BoundLongCounter;
+import io.opentelemetry.metrics.Meter;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.common.DaemonThreadFactory;
+import io.opentelemetry.sdk.common.export.ConfigBuilder;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Logger;
+
+public class BatchLogProcessor implements LogProcessor {
+  private static final String WORKER_THREAD_NAME =
+      BatchLogProcessor.class.getSimpleName() + "_WorkerThread";
+  private static final String TIMER_THREAD_NAME =
+      BatchLogProcessor.class.getSimpleName() + "_TimerThread";
+
+  private final Worker worker;
+  private final Thread workerThread;
+
+  private BatchLogProcessor(
+      int maxQueueSize,
+      long scheduleDelayMillis,
+      int maxExportBatchSize,
+      long exporterTimeout,
+      Exporter exporter) {
+    this.worker =
+        new Worker(
+            maxQueueSize, scheduleDelayMillis, maxExportBatchSize, exporterTimeout, exporter);
+    this.workerThread = new DaemonThreadFactory(WORKER_THREAD_NAME).newThread(worker);
+    this.workerThread.start();
+  }
+
+  public static Builder builder(Exporter exporter) {
+    return new Builder(exporter);
+  }
+
+  @Override
+  public void addLogRecord(LogRecord record) {
+    worker.addLogRecord(record);
+  }
+
+  @Override
+  public void shutdown() {
+    workerThread.interrupt();
+    worker.shutdown();
+  }
+
+  @Override
+  public void forceFlush() {
+    worker.forceFlush();
+  }
+
+  private static class Worker implements Runnable {
+    static {
+      Meter meter = OpenTelemetry.getMeter("io.opentelemetry.sdk.logging");
```

Comment:
We need to be careful to avoid self-reinforcing loops if metric operations decide to log on metering failures like we decide to meter on logging failures here. Do we know that none of the metric API calls we make here is log-free or that metric API implementation will not use our logging methods?

Reply:
This is a good point. I'm also trying to get the message out if logs are *not* getting out- do we have patterns for how we manage this type of situation?

Reply:
The only recommendation I can think of is to be careful and don't make calls that looks mutually recursive. Unfortunately that means you have to do flow control analysis manually. It results in fragile code, I know. Someone touches anything there (adds an innocent looking debug statement) and it can become a disaster :-(

More complicated technical means are possible, like detecting the recursion and aborting it, not sure if we want to go that route.